### PR TITLE
[improve][broker] Reducing the parse of MessageMetadata in compaction

### PR DESCRIPTION
**Pull request by aloyszhang**
### Motivation

There are many redundant parses of MessageMetadata in topic compaction, including phase one and phase two

### Modifications

reuse the MessageMetadata object  in topic compaction

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Documentation

- [x] `doc-not-needed` 

### Matching PR in forked repository

PR in forked repository:  https://github.com/aloyszhang/pulsar/pull/23

### CODE REVIEW

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 1]
```diff
@@ -52,12 +52,14 @@ public static boolean isReadableBatch(MessageMetadata metadata) {
...
 
-    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg)
+    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg, MessageMetadata metadata)
         throws IOException {
         checkArgument(msg.getMessageIdData().getBatchIndex() == -1);
 
         ByteBuf payload = msg.getHeadersAndPayload();
-        MessageMetadata metadata = Commands.parseMessageMetadata(payload);
+        if (metadata == null) {
+            metadata = Commands.parseMessageMetadata(payload);
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:61]: is there a need to skip the message metadata if it's not parsed?
Doesn't `parseMessageMetadata` move the read position of the ByteBuf forward?

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 2]
```diff
@@ -123,7 +135,9 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
             payload.readerIndex(readerIndex);
             brokerMeta = payload.readSlice(brokerEntryMetadataSize + Short.BYTES + Integer.BYTES);
         }
-        MessageMetadata metadata = Commands.parseMessageMetadata(payload);
+        if (metadata == null) {
+            metadata = Commands.parseMessageMetadata(payload);
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:139]: is there a need to skip the message metadata if it's not parsed?
Doesn't `parseMessageMetadata` move the read position of the ByteBuf forward?

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java (hunk 1 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:61]: Yes, you're right. I missed it before.

Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:61]: fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java (hunk 2 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:141]: fixed

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java [hunk 3]
```diff
@@ -52,12 +52,16 @@ public static boolean isReadableBatch(MessageMetadata metadata) {
...
 
-    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg)
+    public static List<MessageCompactionData> extractMessageCompactionData(RawMessage msg, MessageMetadata metadata)
         throws IOException {
         checkArgument(msg.getMessageIdData().getBatchIndex() == -1);
 
         ByteBuf payload = msg.getHeadersAndPayload();
-        MessageMetadata metadata = Commands.parseMessageMetadata(payload);
+        if (metadata == null) {
+            metadata = Commands.parseMessageMetadata(payload);
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:61]: One more concern. `Commands.parseMessageMetadata(payload)` will use a shared thread local instance. Is that a problem here?
`Commands.parseMessageMetadata(payload, new MessageMetadata())` would be a way to resolve it if there's a need to have multiple instances in flight.

Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java:61]: I think it's not a problem.
Metadata is parsed previously when doing compaction in both phase one and phase two.
The null value for metadata only happens in unit tests.

**Comment by lordcheng10** [comment]
LGTM

**Pull request merged**